### PR TITLE
feat: implementing ton-connect 2.0 including js - injected and remote bridges

### PR DESCRIPTION
**Pull request by vzhovnitsky**
This needs a heavy review and probably a lot of refactoring.
https://user-images.githubusercontent.com/32094306/220904414-10bff28e-f6c4-4a63-a948-d0b7552e83cb.MP4

### CODE REVIEW

**Review by dvlkv — CHANGES_REQUESTED** [review]

File: app/engine/tonconnect/fetchManifest.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+export interface AppManifest {
+    url: string;
+    name: string;
+    iconUrl: string;
+    termsOfUseUrl: string | undefined;
+    privacyPolicyUrl: string | undefined;
+}
+
+export async function fetchManifest(link: string) {
```
[the remark this answers is not in the document]
Comment (vzhovnitsky) [reply app/engine/tonconnect/fetchManifest.ts:22]: Indeed, will do!

File: app/fragments/apps/components/ConnectAppComponent.tsx [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    const domain = extractDomain(props.endpoint);
+    const navigation = useTypedNavigation();
+    const start = React.useMemo(() => {
+        return Date.now();
+    }, []);
+    const close = React.useCallback(() => {
+        navigation.goBack();
+        trackEvent(MixpanelEvent.AppClose, { url: props.endpoint, domain, duration: Date.now() - start });
+    }, []);
+    useTrackEvent(MixpanelEvent.AppOpen, { url: props.endpoint, domain });
```
[the remark this answers is not in the document]
Comment (vzhovnitsky) [reply app/fragments/apps/components/ConnectAppComponent.tsx:37]: Yeah, you're right

File: app/engine/tonconnect/TonConnect.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
+import { AppRequest, RpcMethod, SessionCrypto, } from '@tonconnect/protocol';
+import { storage } from '../../storage/storage';
+import { createLogger, warn } from '../../utils/log';
+import { Engine } from '../Engine';
+import { bridgeUrl } from '../products/ConnectProduct';
+import { ConnectedAppConnection, ConnectedAppConnectionRemote, TonConnectBridgeType } from './types';
+import EventSource, { MessageEvent } from 'react-native-sse';
+
+export class TonConnect {
```
Comment (dvlkv) [inline app/engine/tonconnect/TonConnect.ts:9]: TonConnectBridgeClient

File: app/fragments/secure/TonConnectAuthenticateFragment.tsx [hunk 4]
```diff
@@ -0,0 +1,556 @@
...
+                setTimeout(() => {
+                    navigation.goBack();
+                }, 50);
+                return;
+            }
+
+            // Should not happen
+            setState({ type: 'failed' });
+        } catch (error) {
+            warn(error);
```
Comment (dvlkv) [inline app/fragments/secure/TonConnectAuthenticateFragment.tsx:233]: Rewrite to text error to prevent leaking keys

**Review by dvlkv — CHANGES_REQUESTED** [review]

File: app/engine/tonconnect/fetchManifest.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (vzhovnitsky) [reply app/engine/tonconnect/fetchManifest.ts]: Done

File: app/components/TransferBatch.tsx [hunk 5]
```diff
@@ -0,0 +1,612 @@
...
+            return;
+        }
+        if (totalAmount.eq(new BN(0))) {
+            Alert.alert(t('transfer.error.zeroCoins'));
+            return;
+        }
+
+
+        // Read key
+        let walletKeys: WalletKeys;
```
[the remark this answers is not in the document]
Comment (vzhovnitsky) [reply app/components/TransferBatch.tsx:272]: Done

File: app/components/TransferSingle.tsx [hunk 6]
```diff
@@ -0,0 +1,897 @@
...
+        // Read key
+        let walletKeys: WalletKeys;
+        try {
+            walletKeys = await loadWalletKeys(acc.secretKeyEnc);
+        } catch (e) {
+            return;
+        }
+
+        // Create transfer
+        let transfer = await contract.createTransfer({
```
[the remark this answers is not in the document]
Comment (vzhovnitsky) [reply app/components/TransferSingle.tsx]: Done

File: app/engine/tonconnect/TonConnect.ts (hunk 3 above)
Comment (vzhovnitsky) [reply app/engine/tonconnect/TonConnect.ts]: Done

File: app/engine/estimate/estimateFees.ts [hunk 7]
```diff
@@ -2,49 +2,54 @@ import { BN } from "bn.js";
...
-    }) : new BN(0);
-    // console.log('storage_fees: ' + fromNano(storageFees));
+    }
 
     // Calculate import fees
     let importFees = computeExternalMessageFees(config.workchain.message as any, inMsg);
-    // console.log('import_fees: ' + fromNano(importFees));
 
     // Any transaction use this amount of gas
     const gasUsed = 3308;
```
[the remark this answers is not in the document]
Comment (vzhovnitsky) [reply app/engine/estimate/estimateFees.ts]: Done

**Review by dvlkv — APPROVED** [review]
LGTM

**Pull request merged**